oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```diff
@@ -0,0 +1,557 @@
+	 *         are null
+	 * @throws OSException if any of the parameters are unvalid
+	 */
+	public static byte[] nonceGeneration(byte[] partialIV, byte[] senderID, byte[] commonIV, int nonceLength)
+			throws OSException {
+		if (partialIV != null) {
+			if (senderID != null) {
+				if (commonIV != null) {
+					if (nonceLength > 0) {
+						int s = senderID.length;
+						int zeroes = 5 - partialIV.length;
+
+						if (zeroes > 0) {
+							partialIV = leftPaddingZeroes(partialIV, zeroes);
+						}
+
+						zeroes = (nonceLength - 6) - senderID.length;
+
+						if (zeroes > 0) {
+							senderID = leftPaddingZeroes(senderID, zeroes);
+						}
+
+						zeroes = nonceLength - commonIV.length;
+
+						if (zeroes > 0) {
+							commonIV = leftPaddingZeroes(commonIV, zeroes);
+						}
+
+						byte[] tmp = new byte[1 + senderID.length + partialIV.length];
+						tmp[0] = (byte) s;
+						System.arraycopy(senderID, 0, tmp, 1, senderID.length);
+						System.arraycopy(partialIV, 0, tmp, senderID.length + 1, partialIV.length);
+
+						byte[] result = new byte[commonIV.length];
+
+						int i = 0;
+						for (byte b : tmp) {
+							result[i] = (byte) (b ^ commonIV[i++]);
+						}
+
+						return result;
+					} else {
+						LOGGER.severe(OSExceptionHandler.NONCE_LENGTH_INVALID);
+						throw new IllegalArgumentException(OSExceptionHandler.NONCE_LENGTH_INVALID);
+					}
+				} else {
+					LOGGER.severe(OSExceptionHandler.COMMON_IV_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.COMMON_IV_NULL);
+				}
+			} else {
+				LOGGER.severe(OSExceptionHandler.SENDER_ID_NULL);
+				throw new IllegalArgumentException(OSExceptionHandler.SENDER_ID_NULL);
+			}
+		} else {
+			LOGGER.severe(OSExceptionHandler.PARTIAL_IV_NULL);
+			throw new IllegalArgumentException(OSExceptionHandler.PARTIAL_IV_NULL);
+		}
+	}
+	
+	/**
+	 * Padds the left side of the byte array paddMe with zeroes as the int
+	 * zeroes has
+	 * 
+	 * @param paddMe
+	 * @param zeroes
+	 * @return
+	 */
+	private static byte[] leftPaddingZeroes(byte[] paddMe, int zeroes) {
+		byte[] insert = new byte[zeroes];
+		for (int i = 0; i > zeroes; i++) {
+			insert[i] = 0x00;
+		}
+		byte[] tmp = new byte[zeroes + paddMe.length];
+		System.arraycopy(insert, 0, tmp, 0, insert.length);
+		System.arraycopy(paddMe, 0, tmp, insert.length, paddMe.length);
+		return tmp;
+	}
+	
+	/**
+	 * Processes a partialIV correctly
+	 * 
+	 * @param partialIV the partialIV
+	 * @return the processed partialIV
+	 */
+	public static byte[] processPartialIV(int value) {
+		byte[] partialIV = ByteBuffer.allocate(Integer.BYTES).putInt(value).array();
+		return stripZeroes(partialIV);
+	}
+	
+	/**
+	 * Remove trailing zeroes in a byte array
+	 * 
+	 * @param in the incoming array
+	 * @return the array with trailing zeroes removed
+	 */
+	public static byte[] stripZeroes(byte[] in) {
+		if (in != null) {
+			if (in.length == 1)
+				return in;
```

Comment:
So, for length one, nothing is changed.

Reply:
Exactly. We leave one zero behind in the byte array.